support: Add support to configure fixed-price plan with pay-by-invoice.

prakhar1144 opened the pull request:
* Manually create & send invoice
* Configure a fixed-price plan with sent invoice-id.
* When customer pays, upgrade them to concerned plan.


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![image](https://github.com/zulip/zulip/assets/56781761/ecf8040d-20a2-46ef-93f4-db06fe62bcd0)

Customer already on billable plan flow:
* Set end date of current plan
* Change payments method of current plan on /support page
* Configure next fixed-price plan ( Don't send a manual invoice here)

![image](https://github.com/zulip/zulip/assets/56781761/1f4b08d6-0b13-4576-ba82-3626cc2bc026)



<details><summary>After fixed-price plan is configured</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/5a735eb5-554b-47e3-b413-68533dfa13b2">
</img>
</details> 


<details><summary>/upgrade page</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/feb47e3f-e9a2-4220-a959-0a8b2df14f1e">
</img>
</details> 



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

amanagr commented on corporate/lib/stripe_event_handler.py:
```diff
@@ -150,3 +152,35 @@ def handle_payment_intent_succeeded_event(
+            charge_automatically=False,
+            free_trial=False,
+            remote_server_legacy_plan=remote_server_legacy_plan,
+            # TODO: Should we decide for whether to schedule or not via metadata or support ?
+            should_schedule_upgrade_for_legacy_remote_server=remote_server_legacy_plan is not None,
```
Since the customer has already paid, this is a direct upgrade.

amanagr replied:
For scheduled upgrades, we wait till the days of the upgrade to charge customer. So, scheduling an upgrade will cause problems anyway.

prakhar1144 replied:
Ah, that's right! I'll update.

amanagr commented on corporate/views/webhook.py:
```diff
@@ -84,6 +86,17 @@ def stripe_webhook(request: HttpRequest) -> HttpResponse:
         event.object_id = payment_intent.id
         event.save()
         handle_payment_intent_succeeded_event(stripe_payment_intent, event)
+    elif stripe_event.type == "invoice.payment_succeeded":
```
We should also listen for invoice payment failures or what the event for invoices is that we are not able to collect so that we can delete the `CustomerPlan` corresponding to it.

prakhar1144 replied:
We create `CustomerPlan` only after we receive `invoice.payment_succeeded` event.

We create a `CustomerPlanOffer` entry when we configure fixed-price plan via /support. That was one of the motivation behind behind creating this new model. `CustomerPlan` will be clean.

amanagr replied:
Oh! I see.

Also, we should use `invoice.paid` event here.

amanagr replied:
https://stripe.com/docs/api/events/types#event_types-invoice.paid

prakhar1144 commented on corporate/lib/stripe.py:
```diff
@@ -1412,8 +1422,10 @@ def process_initial_upgrade(
+        if charge_automatically:
+            customer = self.update_or_create_stripe_customer()
+        else:
+            customer = self.update_or_create_customer()
```
> Since we are no longer creating an invoice here for non charge_automatically plans

I took a slightly different approach in the updated code. Now, a boolean `stripe_invoice_payment_succeeded` will decide whether to create an invoice or not.

We still have code for cloud customers where for "Pay by invoice" billing modality, it immediately upgrades them and generates an invoice. This is not possible by UI. But this would be a separate project altogether to decide how to deal with customers who want to upgrade to various plans with "Pay by invoice".

So, the current code offers three possibility:
* Charge automatically
* Manually send invoice for fixed price plan & customer automatically upgrades as soon as they pay.
* Choose billing_modality = "Pay by invoice", we upgrade instantly & invoice is sent. (Backend code is available but no way to select "Pay by invoice" in /upgrade.)

timabbott commented on corporate/views/support.py:
```diff
@@ -556,6 +553,7 @@ def remote_servers_support(
             support_view_request = SupportViewRequest(
                 support_type=SupportType.configure_fixed_price_plan,
                 fixed_price=fixed_price,
+                sent_invoice_id=sent_invoice_id,
```
Does this end up supporting invoices in both full Stripe URL and ID format? We might want both to work?

amanagr replied:
Oh, yeah, that's a good idea. We just need to extract the ID from the URL.

prakhar1144 replied:
[CZO Discussion](https://chat.zulip.org/#narrow/stream/191-kandra-support/topic/invoice-based.20billing/near/1731749)

We make an API call to fetch payments-page URL on /upgrade. API call is made when a customer tries to visit /upgrade after fixed-price offer is configured with sent_stripe_id, so this is going to be rare.

prakhar1144 commented:
Updates in the recent push:
* Added first commit -- Explanation in commit message.
* Added tests
* /upgrade page -- [CZO Discussion](https://chat.zulip.org/#narrow/stream/191-kandra-support/topic/invoice-based.20billing/near/1730896):  "it's probably best to block upgrading via /upgrade (maybe with a link to the invoice)"

timabbott commented on web/src/billing/upgrade.ts:
```diff
@@ -29,6 +29,10 @@ if ($("input[type=hidden][name=schedule]").length === 1) {
+if (page_params.fixed_price !== null) {
+    // By default, we show annual schedule (and price) for a fixed-price plan.
+    selected_schedule = "annual";
+}
```
Do we want this to override what was in local storage? Maybe it doesn't matter but it seems like the intent would be that we should just replace the default in the line `typeof ls_selected_schedule === "string" ? ls_selected_schedule : "monthly";` with defining a `default_selected_schedule` based on the plan type. Hmm, I guess that approach might be annoying if you went to /upgrade before your fixed-price plan code programmed. I think this is probably OK for now given that analysis.

timabbott commented on templates/corporate/support/next_plan_forms_support.html:
```diff
@@ -1,8 +1,21 @@
+  current_plan.TIER_SELF_HOSTED_BASIC,
+  current_plan.TIER_SELF_HOSTED_BUSINESS
+  ]
+  ) %}
```
We shouldn't have a list of plans hardcoded in template files; we should pass the determination as to whether the current plan is billable in from the caller. And probably that code should call `check_plan_tier_is_billable`.

prakhar1144 replied:
Made the changes.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1291,6 +1294,14 @@ def configure_fixed_price_plan(self, fixed_price: int) -> str:
+                customer=customer,
+                stripe_invoice_id=sent_invoice_id.strip(),
+                status=Invoice.SENT,
+            )
```
It could be good to, before creating the invoice object, query the Stripe API to ask whether it can find the invoice ID, just an end-to-end check against typos before storing it in the database.

prakhar1144 replied:
![image](https://github.com/zulip/zulip/assets/56781761/37056e93-1911-4acf-bd8a-f8f6412c4661)

Explicitly added this 'status=open' check, because I expect there's a good chance of handling multiple support request with a lot of tabs open, resulting in copy-pasting of paid/draft invoice ID.
![image](https://github.com/zulip/zulip/assets/56781761/7878ba51-6d25-4dd0-9785-6a570957c244)

prakhar1144 replied:
I think removing the "Request req_..." is not of much importance as this is /support page. The error is returned by stripe, which would vary depending on error. I can improve this error message presentation, if required.

timabbott replied:
Yeah that sounds right to me.

timabbott commented on corporate/lib/stripe_event_handler.py:
```diff
@@ -150,3 +152,34 @@ def handle_payment_intent_succeeded_event(
+    if customer.required_plan_tier:
+        configured_fixed_price_plan = get_configured_fixed_price_plan_offer(
+            customer, customer.required_plan_tier
+        )
```
This logic is fine for now, but I think we should still be planning to move `required_plan_tier` to be a `tier` field of `CustomerPlanOffer`.

prakhar1144 replied:
Yeah, planning to do this next to prevent more such additions.

timabbott commented on corporate/lib/stripe_event_handler.py:
```diff
@@ -150,3 +152,34 @@ def handle_payment_intent_succeeded_event(
+            customer, customer.required_plan_tier
+        )
+
+    if stripe_invoice.collection_method == "send_invoice" and configured_fixed_price_plan:
```
Do we want/need to do logging when this function would exit without running this code? I wonder if we want to add some logging/assertions here.

prakhar1144 replied:
Actually, this function will be extended to handle other cases when this **if** condition is not met. That would be the part of the work Aman is doing (#28786) to [replace PaymentIntent with Invoice](https://chat.zulip.org/#narrow/stream/273-kandra-billing/topic/.240.20upgrade.20invoices/near/1729825) only.

So, I think, whatever logs/assertion we'd end up with will be finalized in #28786 ?

timabbott replied:
OK that sounds reasonable to me. @amanagr FYI.

timabbott commented:
This looks great! I posted a handful of comments about ways to make this a bit more robust, but the implementation looks solid.

prakhar1144 commented:
Addressed the review comments. -- replied to the respective review comments.

alya commented:
As a tiny comment, the spacing between paragraphs on the /upgrade screenshot looks surprisingly large to me.

timabbott commented:
Looks great, merged, thanks @prakhar1144! @amanagr it is probably worth your looking at the spacing issue that @alya noted.

amanagr commented:
Opened #29000
